### DISCUSSION TITLE
Publishin data to clients using milo vs http

### DISCUSSION THREAD

**Discussion by ronytesler** [eclipse-milo/milo#1020] [General]
We have an application 'A' that connects to some opcua servers. 'A' has its own clients that need to get the data from 'A'.
Different clients might need data from the same opcua servers, but with different publish time intervals.
What would be the best way to implement the communication between 'A' and its clients?
- Using milo to make 'A' an opcua server as well, and updating its nodes with new data every time data is coming from the "real" opcua servers. 
- Calling each client using http, passing the relevant data, using a timer for sending it in the correct time.
- Something else?

**Comment by kevinherron** [comment]
"Best" is up to you. Can you modify the clients of "A" to be OPC UA clients? Do you have the resources to turn "A" into a server as well?

The HTTP (or maybe WebSocket) approach would work fine too. The clients would call an endpoint to set up a corresponding subscription with the desired publishing interval, etc... and then just poll or wait for data over a WebSocket.

**Reply by ronytesler** [reply]
I don't want the clients to poll because maybe there won't be a new data for a long time so there will be no need in these calls. I want the server to know when to call each client.
I can turn the clients of 'A' to be opc ua clients. About the resources to turn 'A' into an opc ua server - what are the needed resources, especially compared to what's needed for the http (or WebSocket) implementation?

**Reply by kevinherron** [reply]
I don't really mean computing resources, just developer resources. Developing an OPC UA server can be a bit more involved than a client and the server SDK isn't as well documented.

**Reply by ronytesler** [reply]
Ah I understand. Maybe we already have an implementation of a server so maybe that can save time. Should we consider using other library where its server side is easier to use?

**Reply by kevinherron** [reply]
Up to you. If it's in Java you don't really have any other options unless you want to pay for Prosys, and even then it's still involved.

**Reply by ronytesler** [reply]
Thanks!!
